modules/standard/FileSystem.chpl

```diff
@@ -302,79 +283,136 @@ proc copy(out error: syserr, src: string, dest: string, metadata: bool = false)
+   by `dest`, replacing `dest`'s contents if it already exists (and is a
+   different file than `src`, i.e. not a symbolic link to `src`).
+
+   If `dest` is not writable, or `src` and `dest` refer to the same file,
+   this function will throw an error.  Does not copy metadata.  May
+   throw other error messages.
+
+   :arg src: The source file whose contents are to be copied.
+   :type src: `string`
+   :arg dest: The intended destination of the contents.
+   :type dest: `string`
+*/
+proc copyFile(src: string, dest: string) throws {
   // This implementation is based off of the python implementation for copyfile,
   // with some slight differences.  That implementation was found at:
   // https://bitbucket.org/mirror/cpython/src/c8ce5bca0fcda4307f7ac5d69103ce128a562705/Lib/shutil.py?at=default
   // I did not look at the other functions in that file, except for copyfileobj
   // (which copyfile called).
-  var exist = exists(error=error, src);
-  if error then
-    return;
 
-  if (!exist) {
-    error = ENOENT;
+  if !(try exists(src)) then
     // Source didn't exist, we can't copy it.
-    return;
-  }
-  if (isDir(error, src) || isDir(error, dest)) {
-    // If the source is a directory, the user has made a mistake, so return an
-    // error.  The same is true if the destination is a directory.
-    error = EISDIR;
-    return;
-  }
+    try ioerror(ENOENT:syserr, "in copyFile(" + src + ", " + dest + ")");
 
-  if (error == ENOENT) {
-    error = ENOERR;
-    // We don't care if dest did not exist before, we'll create or overwrite it
-    // anyways.  We already know src exists.
-  } else if (sameFile(error, src, dest)) {
-    // Check if the files are the same, error if yes
+  try {
+    if (isDir(src) || isDir(dest)) {
+      // If the source is a directory, the user has made a mistake, so return
+      // an error.  The same is true if the destination is a directory.
+      ioerror(EISDIR:syserr, "in copyFile(" + src + ", " + dest + ")");
+    }
 
-    // Don't need to check if they're the same file when we know dest didn't
-    // exist.
-    error = EINVAL;
-    return;
-    // The second argument is invalid if the two arguments are the same.
+    if (sameFile(src, dest)) {
+      // Check if the files are the same, error if yes
+      try ioerror(EINVAL:syserr, "in copyFile(" + src + ", " + dest + ")");
+    }
+  } catch e: FileNotFoundError {
+    // We don't care if dest did not exist before, we'll create or overwrite
+    // it anyways.  We already know src exists.
   }
 
   // Open src for reading, open dest for writing
-  var srcFile = open(src, iomode.r, error=error);
-  if error then return;
-  var destFile = open(dest, iomode.cw, error=error);
-  if error {
-    var ignore:syserr;
-    srcFile.close(error=ignore);
-    return;
+  var srcFile = try open(src, iomode.r);
+  defer {
+    try {
+      srcFile.close();
+    } catch {
+      // ignore errors
```

Comment:
As with IO functions, we shouldn't be ignoring an error on a close. We can ignore an error on close if it happens in the process of returning an error. But if everything else ran OK, an error on close should cause the function to throw. This pattern occurs several times in this function.